File: website/functions/_lib/http.ts
/**
 * 统一 JSON 响应、错误信封与输入解析（web-design.md §8.2）。
 *
 * 错误信封形状（全端点一致）：
 *   { "error": { "code": "...", "message": "..." } }
 * 不回显 token / secret / 堆栈。
 */

const JSON_HEADERS: Record<string, string> = {
  "content-type": "application/json; charset=utf-8",
  // 端点不缓存（含敏感响应体）
  "cache-control": "no-store",
};

/** 成功 JSON 响应。 */
export function json(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), { status, headers: JSON_HEADERS });
}

/** 错误信封响应。 */
export function jsonError(status: number, code: string, message: string): Response {
  return json(status, { error: { code, message } });
}

export function badRequest(message: string, code = "BAD_REQUEST"): Response {
  return jsonError(400, code, message);
}

export function notFound(code = "NOT_FOUND", message = "resource not found"): Response {
  return jsonError(404, code, message);
}

export function methodNotAllowed(): Response {
  return jsonError(405, "METHOD_NOT_ALLOWED", "method not allowed");
}

/** 安全解析 JSON body；非法/超大返回 null（由调用方转 400）。 */
export async function readJson<T = Record<string, unknown>>(
  request: Request,
  maxBytes = 8 * 1024,
): Promise<T | null> {
  const contentType = request.headers.get("content-type") ?? "";
  if (!contentType.toLowerCase().includes("application/json")) return null;
  const raw = await request.text();
  if (raw.length === 0 || raw.length > maxBytes) return null;
  try {
    return JSON.parse(raw) as T;
  } catch {
    return null;
  }
}

/**
 * expires_in 归一化（web-design.md §5 端点3/4 约定）：
 * 必须为正整数秒；IdP 未返回/≤0/非法 → 回退保守默认。
 */
export function normalizeExpiresIn(value: unknown, fallbackSeconds = 300): number {
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(n)) return fallbackSeconds;
  const floored = Math.floor(n);
  return floored > 0 ? floored : fallbackSeconds;
}

/** 从对象读必填字符串字段；缺失/非字符串/超长返回 null。 */
export function readStringField(
  body: Record<string, unknown> | null,
  key: string,
  maxLen = 4096,
): string | null {
  if (!body) return null;
  const v = body[key];
  if (typeof v !== "string" || v.length === 0 || v.length > maxLen) return null;
  return v;
}

File: server/internal/broker/http.go
package broker

import (
	"encoding/json"
	"io"
	"math"
	"net/http"
	"strconv"
	"strings"
)

// The broker uses its own JSON envelope, distinct from rest's
// {success,data,error} (decision 10). Every endpoint returns either a bare
// success body or the error shape below (mirrors _lib/http.ts):
//
//	{ "error": { "code": "...", "message": "..." } }
//
// Responses never echo tokens, secrets, or stack traces.

// brokerError is the error envelope payload (mirrors http.ts jsonError).
type brokerError struct {
	Error errorBody `json:"error"`
}

type errorBody struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// defaultReadJSONMaxBytes bounds request bodies (mirrors http.ts readJson's
// 8 KiB default). Anything larger is treated as a bad request by the caller.
const defaultReadJSONMaxBytes = 8 * 1024

// defaultExpiresIn is the conservative fallback when the IdP omits or returns an
// invalid expires_in (mirrors http.ts normalizeExpiresIn fallbackSeconds=300).
const defaultExpiresIn = 300

// writeJSON writes v as a JSON body with the given status. Endpoints are never
// cached (they carry sensitive bodies), mirroring http.ts JSON_HEADERS.
func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

// writeError writes the {error:{code,message}} envelope (mirrors http.ts
// jsonError).
func writeError(w http.ResponseWriter, status int, code, message string) {
	writeJSON(w, status, brokerError{Error: errorBody{Code: code, Message: message}})
}

// badRequest writes a 400 with the given message and code (default BAD_REQUEST),
// mirroring http.ts badRequest.
func badRequest(w http.ResponseWriter, message string, code ...string) {
	c := "BAD_REQUEST"
	if len(code) > 0 && code[0] != "" {
		c = code[0]
	}
	writeError(w, http.StatusBadRequest, c, message)
}

// notFound writes a 404 with the given code/message (mirrors http.ts notFound).
func notFound(w http.ResponseWriter, code, message string) {
	writeError(w, http.StatusNotFound, code, message)
}

// readJSON safely decodes a JSON object body into a map, returning (nil, false)
// on a non-JSON content type, an empty/oversized body, or a parse error
// (mirrors http.ts readJson: the caller turns a nil result into a 400). It never
// reads more than maxBytes.
func readJSON(r *http.Request, maxBytes int64) (map[string]any, bool) {
	if maxBytes <= 0 {
		maxBytes = defaultReadJSONMaxBytes
	}
	ct := strings.ToLower(r.Header.Get("Content-Type"))
	if !strings.Contains(ct, "application/json") {
		return nil, false
	}
	// Read up to maxBytes+1 so an over-limit body is detected without buffering
	// an unbounded amount (mirrors the raw.length > maxBytes check in http.ts).
	raw, err := io.ReadAll(io.LimitReader(r.Body, maxBytes+1))
	if err != nil {
		return nil, false
	}
	if len(raw) == 0 || int64(len(raw)) > maxBytes {
		return nil, false
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil || out == nil {
		return nil, false
	}
	return out, true
}

// readStringField reads a required non-empty string field, rejecting missing,
// non-string, empty, or over-long values (mirrors http.ts readStringField). It
// returns ("", false) so callers map failure to a 400.
func readStringField(body map[string]any, key string, maxLen int) (string, bool) {
	if body == nil {
		return "", false
	}
	if maxLen <= 0 {
		maxLen = 4096
	}
	v, ok := body[key].(string)
	if !ok || v == "" || len(v) > maxLen {
		return "", false
	}
	return v, true
}

// normalizeExpiresIn coerces an IdP expires_in to a positive integer number of
// seconds, falling back to defaultExpiresIn on absent/invalid/<=0 values
// (mirrors http.ts normalizeExpiresIn). JSON numbers decode as float64.
func normalizeExpiresIn(value any) int {
	var n float64
	switch v := value.(type) {
	case float64:
		n = v
	case int:
		n = float64(v)
	case int64:
		n = float64(v)
	case string:
		// Some IdPs return expires_in as a JSON string (e.g. "3600"); mirror
		// the TS Number() coercion. Unparseable -> fallback.
		f, err := strconv.ParseFloat(v, 64)
		if err != nil {
			return defaultExpiresIn
		}
		n = f
	default:
		return defaultExpiresIn
	}
	if math.IsNaN(n) || math.IsInf(n, 0) {
		return defaultExpiresIn
	}
	floored := int(math.Floor(n))
	if floored > 0 {
		return floored
	}
	return defaultExpiresIn
}
